WIP: remove perl build dependency

ioraff opened the pull request.

ioraff committed:
remove perl build dep

brada4 commented:
Did not know bash-isms do work on Windoze

martin-frbg commented:
> Did not know bash-isms do work on Windoze

Probably requires either WSL or a native installation of git (for gitbash) ? (Not sure how much of an improvement over requiring perl it would be, also some Unix systems may not come with bash installed. Doing it all in a minimal POSIX sh seemed impossible last I looked into this)

ioraff commented:
> Did not know bash-isms do work on Windoze

There aren't any bashisms here

martin-frbg commented:
Impressive then. As you can tell I have not had time to look at it in detail, but the crashes in CI suggest there may be a problem with detecting the right representation of complex variables on some platforms. (There are some errors in tests like SASUM that I do not understand yet, may make sense to compare the configuration files created by the perl and sh versions in each case. And f_check line 95 seems to have a problem with obtaining the version number of the fortran compiler in some cases)

martin-frbg commented:
for f_check around line 95 I suspect the pattern matching should start with `v=${data##*GCC: *\) }` to catch all the vendor/distributor compilers that call themselves "GCC: (vendorname-somepattern) X.Y.Z" rather than the canonic "GCC: (GNU) X.Y.Z"

ioraff committed:
attempt to fix GNU fortran detection, fix definition printing

ioraff committed:
readlink -f is non-POSIX

ioraff committed:
attempt to fix makefile generation

ioraff committed:
parentheses match

ioraff committed:
gensymbol: fix type

ioraff committed:
OS potentially being overwritten

brada4 commented on c_check:
```diff
@@ -1,427 +1,381 @@
-    $openmp = "-fopenmp";
+{
+    data=$($compiler_name $flags -E ctest.c)
+} || {
```
Please use if [ everywhere or shorthands everywhere.

martin-frbg replied:
brada4 I am not convinced that unification would improve readability, both variants have their uses

brada4 replied:
this looks well... obfuscated, more verbose version as everywhere else would be more handy

ioraff replied:
To my knowledge, the way I've used both variants here is pretty standard practice in shell scripts. Rewriting [this](https://github.com/xianyi/OpenBLAS/blob/7c146da155d628de02f17c4ffafb05c2af93a868/f_check#L69), for example, with only if-else would be a mess.

martin-frbg replied:
To be clear I am happy with merging this as-is. I only want to try it on some machines not available in the current CI setup first (and maybe keep the perl scripts around somewhere as a fallback, but that can be looked into later)

ioraff committed:
clean and fix c_check makefile generation

mmuetzel committed:
Consolidate actions on GitHub runners.

Re-organize build matrix for Ubuntu and MacOS runners.
Don't start runners that don't do anything.
Run tests.

martin-frbg committed:
Use f2c translations of LAPACK when no Fortran compiler is available (#3539)

* Add C equivalents of the Fortran routines from Reference-LAPACK as fallbacks, and C_LAPACK variable to trigger their use

mmuetzel committed:
Adapt commands for tests with GNU make.

martin-frbg committed:
fix undefined prefetchsize

martin-frbg committed:
fix undefined prefetchsizes

martin-frbg committed:
C_LAPACK: Fixes to make it compile with MSVC (#3605)

* Fix f2c-like support functions to compile with MSVC, and
re-enable C_LAPACK for MSVC in CMAKE

* Add MSVC&flang build to Azure CI in order to check C_LAPACK correctness

martin-frbg commented:
quick update, sorry for the delay - Solaris sh does not like the `f="${flag#\}"` near line 295 of c_check (gives "bad substitution")
and it also looks as if the contents of the "old_flags" list has been replaced with the verbatim "echo" command at this point, rather than its output

martin-frbg commented:
looks like all `$( ... )` command substitutions need to be replaced with backticks syntax as well

martin-frbg commented:
(this is with Solaris 10 to be precise - host gcc210 in the GCC compile farm.)

niyas-sait committed:
build: minor fixes to build on windows with make

This patch contains following fixes

1. Fix to build without PIC flag
2. Define LAPACK_COMPLEX_STRUCTURE for windows. Builds are failing
without it and changes are consistent with the CMake rules defined
in system.cmake (line 576)

brada4 commented:
It fails on "common linux" too, say dash on Debian:
```
exports/gensymbol: 3620: [: Illegal number: 
c_check: 21: shift: can't shift that many
```
I dont think it is practical to introduce new bash dependency on !linux platforms.

martin-frbg commented:
Nobody plans to replace the perl dependency with a bash one, the question is simply if it is possible to rewrite the script in an idiom that is understood by all shells (and apparently POSIX is asking too much here, seems the old Suns use the original Bourne shell).

brada4 commented:
Indeed, probably worth documenting what config files get generated there so that one willing can get through without bash or perl.

ioraff commented:
> It fails on "common linux" too, say dash on Debian:
> 
> ```
> exports/gensymbol: 3620: [: Illegal number: 
> c_check: 21: shift: can't shift that many
> ```
> 
> I dont think it is practical to introduce new bash dependency on !linux platforms.

Again, there aren't any bashisms here. Perhaps that was a bug with an older dash, as I have no problems running this with dash 0.5.11.5.

martin-frbg commented:
Still a WIP, but this appears to work with both SunOS' bourne shell and Linux' bash
[shellfiles.zip](https://github.com/xianyi/OpenBLAS/files/8557721/shellfiles.zip)

ioraff committed:
use backticks

martin-frbg committed:
Initial attempt at proper cpu detection on RISCV

martin-frbg committed:
Initialize oldgcc,no_avx2  and add the rv64gv check

martin-frbg commented:
Sorry, it seems I cannot resolve the merge conflict (that I probably created a week ago with the riscv additions to the perl c_check) from my side. Can you please try if you can override it with a force-push again ?

ioraff committed:
c_check: fix syntax

ioraff commented:
well, I've clearly broken this. not sure what to do now.

martin-frbg commented:
hmm. my low-tech idea would be to create a new branch from current `develop`, copy the changed files from the stalled branch over and do a new PR ?

martin-frbg commented:
(I'm not sure I understand what exactly went wrong, except that a force-push was probably a very bad suggestion). I may still have a .diff version of this PR from happier times

martin-frbg commented:
This is what it looked like on April 10 (which is what I based my Apr 24 comments on - https://github.com/xianyi/OpenBLAS/pull/3578#issuecomment-1107876155
[3578.diff.gz](https://github.com/xianyi/OpenBLAS/files/8681548/3578.diff.gz)
and it looks like the links to your individual changes since then ("readlink -f is not POSIX" et al.) stlll work (and so does the usual github magic whereby appending ".diff" to the URL of the commit view turns it into a raw context diff suitable for patching)

martin-frbg commented:
I have now checked that all individual commits preceding "use backticks" are already contained in the 3578.diff of April 10, so only this one and my attempt to update your c_check today (minus the typos) would need to be reapplied on top of it

ioraff commented:
A force-push was probably the right idea. I just borked the rebase.

The diff fails to apply to the current `develop` `HEAD` for me. Not sure how to go about debugging it since the changes are basically all in one hunk. Might just rewrite at this point.

martin-frbg commented:
Hmm, I'm seeing a patch failure in Makefile.prebuild, and the major one in c_check. Easiest solution for the latter would probably be to check out an earlier stage of `develop`, i.e. from April 10 or thereabouts, apply the patch there and copy the c_check file into the new PR branch

martin-frbg commented:
This here is from applying 3578.diff to an April 17 checkout of `develop` (18b19d13 , or right after merging #3605)
[c_check.txt](https://github.com/xianyi/OpenBLAS/files/8682280/c_check.txt)
(and the diff for Makefile.prebuild was only about removing the "perl" command from script invocations)

brada4 commented:
> 0.5.11.5

It has to work on existing platforms, including RHEL6, SLES 12, DEBIAN 9, all BSD-s etc.
